planner/funcdep/fd_graph.go

```diff
@@ -744,47 +838,88 @@ func (s *FDSet) MakeOuterJoin(innerFDs, filterFDs *FDSet, outerCols, innerCols F
 		// This is because that the outer join would generate null-extended rows. So if at least one row from the left side
 		// is not null. We can guarantee that the there's no same part between the original rows and the generated rows.
 		// So the null extended rows would not break the original functional dependency.
-		if edge.from.Intersects(innerFDs.NotNullCols) {
+		if edge.from.Intersects(innerFDs.NotNullCols.Union(filterFDs.NotNullCols)) {
 			// One of determinant are not null column, strict FD are kept.
-			// According knowledge #2, we can't take use of right filter's not null attribute.
 			s.addFunctionalDependency(edge.from, edge.to, edge.strict, edge.equiv)
 		} else {
 			// Otherwise, the strict FD are downgraded to a lax one.
 			s.addFunctionalDependency(edge.from, edge.to, false, edge.equiv)
 		}
 	}
+	// handle the ncEdges from inner side before innerFD, because the current join predicate may bring the visibility to some
+	// cond-fds from the inner side. (which should be affected by null-supplied row)
+	// 1: equivalence cond-fd: once the join predicate can bring it visible again, the equivalence FD exists regardless of null-supplied row.
+	// 2: constant null fd: once the join predicate can bring it visible again, only the part of constant null FD exists regardless of null-supplied row.
+	for i := 0; i < len(innerFDs.ncEdges); i++ {
```

Comment:
We can use
```
for i := len(arr); i >= 0; i-- {
    if satisfied {
        arr = append(arr[:i], arr[i+1:]...)
    }
}
```
to avoid the `i--` at line#858